src/librustc/ty/maps/plumbing.rs

```diff
@@ -277,10 +252,47 @@ macro_rules! define_maps {
                 DepNode::new(tcx, $node(*key))
             }
 
+            fn try_get_lock(tcx: TyCtxt<'a, $tcx, 'lcx>,
+                            mut span: Span,
+                            key: &$K)
+                            -> Result<LockGuard<'a, QueryMap<$tcx, Self>>,
```

Comment:
Can we make a custom enum here? This type is sort of incomprehensible. Something with named variants seems better. Like:

```rust
enum TryGetLock {
    NotYetStarted(Guard), // Today: Ok(guard)
    JobCompleted(...), // Today: Err(Ok(...))
    CycleDetected(...), // Today: Err(Err...))
}
```

I guess it means:

- `Ok(guard)` -- no job yet started, lock is still held
- `Err(Ok(...))` -- cache hit, or else we blocked on a job and it completed
- `Err(Err(..))` -- cycle detected